kafka/src/main/java/com/linecorp/armeria/server/logging/structured/KafkaServiceLogWriter.java

```diff
@@ -0,0 +1,99 @@
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.logging.ResponseLog;
+
+/**
+ * Kafka backend for service log logging.
+ * This class enable Kafka as a service log logging backend.
+ * As a typical sue, you can give constructed Kafka {@link Producer} and the topic which the service logs
+ * will be stored.
+ * {@code
+ * Producer<byte[], ApacheThriftServiceLogEntry> producer = new KafkaProducer<>(...);
+ * ServiceLogWriter<ApacheThriftServiceLogEntry> writer =
+ *     new KafkaServiceLogWriter<>(producer, "airemra-service-log");
+ * }
+ * You may want to supply semantically meaningful "key" to let Kafka(for example) to partition logs in order:
+ * {@code
+ * ServiceLogWriter<ApacheThriftServiceLogEntry> writer =
+ *     new KafkaServiceLogWriter<>(producer, "airemra-service-log",
+ *                                 (ctx, res, entry) ->
+ *                                     String userId = ((HttpRequest) ctx.request()).headers()
+ *                                         .get(AsciiString.of("X-User-Id").getBytes());
+ * }
+ */
+public class KafkaServiceLogWriter<E> implements ServiceLogWriter<E> {
+    private static final Logger logger = LoggerFactory.getLogger(KafkaServiceLogWriter.class);
+
+    /**
+     * Implements "key" selector of Kafka based service log writer.
+     * Kafka as a notion of the "key" which is used as a criteria to guarantee message ordering and
+     * to achieve fine-distributed message partitioning.
+     * Users needs to implement this interface in order to select arbitrary key from the request context
+     * or from the content included in a request.
+     */
+    @FunctionalInterface
+    public interface KeySelector<E> {
+        /**
+         * Selects a key which should be associated toe the record given as {@code entry}
+         * @return A byte-array represented key or null
+         */
+        byte[] selectKey(RequestContext reqCtx, ResponseLog resLog, E entry);
+    }
+
+    private final Producer<byte[], E> producer;
+    private final String topic;
+    private final KeySelector<E> keySelector;
+
+    /**
+     * Constructs {@link KafkaServiceLogWriter} with fully customized arguments.
+     * @param producer a kafka {@link Producer} producer which is used to send logs to Kafka
+     * @param topic a name of topic which is used to send logs
+     * @param keySelector a {@link KeySelector} which is used to decide what key to use for the log
+     */
+    public KafkaServiceLogWriter(Producer<byte[], E> producer, String topic, KeySelector<E> keySelector) {
+        this.producer = producer;
+        this.topic = topic;
+        this.keySelector = keySelector;
+    }
+
+    /**
+     * Constructs {@link KafkaServiceLogWriter}.
+     */
+    public KafkaServiceLogWriter(Producer<byte[], E> producer, String topic) {
```

Comment:
You can get notified when the server shuts down by adding a listener to it:

https://github.com/line/armeria/blob/master/jetty/src/main/java/com/linecorp/armeria/server/http/jetty/JettyService.java#L175